src/core/expression/qgsexpressionnodeimpl.cpp

```diff
@@ -298,14 +298,22 @@ QVariant QgsExpressionNodeBinaryOperator::evalNode( QgsExpression *parent, const
 
     case boAnd:
     {
-      QgsExpressionUtils::TVL tvlL = QgsExpressionUtils::getTVLValue( vL, parent ), tvlR = QgsExpressionUtils::getTVLValue( vR, parent );
+      QgsExpressionUtils::TVL tvlL = QgsExpressionUtils::getTVLValue( vL, parent );
+      ENSURE_NO_EVAL_ERROR;
+      if ( tvlL == QgsExpressionUtils::False )
+        return TVL_False;  // shortcut -- no need to evaluate right-hand side
```

Comment:
Could the same shortcut apply for nulls?

Reply:
Unfortunately not if we want to be 100% correct with three-value boolean logic. If left value is null/unknown, there are still two options: if right-hand side is FALSE, the result will be FALSE, or null/unknown otherwise.